### ISSUE TITLE
Cannot update ConfigRepo through API

### ISSUE THREAD

**Issue by maheshp** [gocd/gocd#5394]
##### Issue Type

- Bug Report

##### Summary

Updates of config repo through [API](https://api.gocd.org/current/#update-config-repo) does not work. The json returned is same as before update.

###### Basic environment details

* Go Version: `18.11.0 - latest experimental`

**Comment by rajiesh** [comment]
Issues in config repo to be fixed for 18.11 release:

- Appropriate error message not provided in response JSON if same material url is used in multiple config repo,

Create config `repo1` using json,

```{
    "id": "repo1",
    "plugin_id": "json.config.plugin",
    "material": {
      "type": "git",
      "attributes": {
        "url": "http://github.com/test-project",
        "auto_update": true,
        "branch": "master"
      }
    }
  }
```

Create config `repo2` using json(the material url is same as repo1)

```{
    "id": "repo2",
    "plugin_id": "json.config.plugin",
    "material": {
      "type": "git",
      "attributes": {
        "url": "http://github.com/test-project",
        "auto_update": true,
        "branch": "master"
      }
    }
  }
```

Second request fails with response as below, response do not contain the appropriate error message

```{
    "message": "Validations failed for config-repo 'repo2'. Error(s): [Validation failed.]. Please correct and resubmit.",
    "data": {
        "id": "repo2",
        "plugin_id": "json.config.plugin",
        "material": {
            "type": "git",
            "attributes": {
                "name": null,
                "auto_update": true,
                "url": "http://github.com/test-project",
                "branch": "master"
            }
        },
        "configuration": []
    }
}
```

- Invalid material type returns 500 server error,

Submit a create config repo request as below,

```
{
    "id": "repo2",
    "plugin_id": "json.config.plugin",
    "material": {
      "type": "GIT",
      "attributes": {
        "url": "http://github.com/test-project",
        "auto_update": true,
        "branch": "master"
      }
    }
  }
```

API returns 500 server error. Error in logs,

```
2018-11-14 16:24:13,091 ERROR [qtp84739718-14] GeneralError:54 -
java.lang.IllegalArgumentException: Unsupported material type: GIT
	at com.thoughtworks.go.apiv1.configrepos.representers.MaterialRepresenter.fromJSON(MaterialRepresenter.java:114)
	at com.thoughtworks.go.apiv1.configrepos.representers.ConfigRepoConfigRepresenterV1.fromJSON(ConfigRepoConfigRepresenterV1.java:37)
	at com.thoughtworks.go.apiv1.configrepos.ConfigReposControllerV1.buildEntityFromRequestBody(ConfigReposControllerV1.java:163)
	at com.thoughtworks.go.apiv1.configrepos.ConfigReposControllerV1.createRepo(ConfigReposControllerV1.java:117)
	at spark.RouteImpl$1.handle(RouteImpl.java:72)
	at spark.http.matching.Routes.execute(Routes.java:61)
	at spark.http.matching.MatcherFilter.doFilter(MatcherFilter.java:130)
	at spark.servlet.SparkFilter.doFilter(SparkFilter.java:173)
	at com.thoughtworks.go.spark.SparkPreFilter.doFilter(SparkPreFilter.java:53)
```

- Plugin id is not validated(looks like this issue always existed with the API, It can be discussed if need to be fixed in 18.11 release itself)

Submit a create or update config repo API call with below request,

```
{
    "id": "repo1",
    "plugin_id": "foo.bar",
    "material": {
      "type": "git",
      "attributes": {
        "url": "http://github.com/test-project",
        "auto_update": true,
        "branch": "master"
      }
    }
  }
```

Repo created successfully in config xml

**Comment by adityasood** [comment]
Verified the error message for duplicate Config-repo URL and all CRUD operations for Config-repo API on 18.11.0-7990.

Moving the following issues 

1. Invalid material type returns 500 server error
2. Plugin id is not validated

To a placeholder issue https://github.com/gocd/gocd/issues/5430, to be looked at as part of a later release.

Closing this issue.
